### PR TITLE
constexpr type ids

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26628 [wip] Remove CAFFE_KNOWN_TYPE
* #26619 [wip] caffe2::TypeMeta uses compile time type names
* #26618 [wip] Compile time type names
* #26616 c10::string_view
* #26509 Remove preallocation of type ids
* **#26502 constexpr type ids**

Create type ids at compile time instead of incrementing a counter at runtime. This is done by computing a compile time crc64 on the type name. We couldn't do this before, because we still used GCC4 and that compiler didn't support the use of __PRETTY_FUNCTION__ in a constexpr context. However, since GCC5 this is possible and we can use this trick.

This does not change the semantics of preallocated type ids. I actually think we don't need to preallocate anymore, but I split the removal of preallocation into a separate diff to be able to test it separately.

Differential Revision: [D17488861](https://our.internmc.facebook.com/intern/diff/D17488861/)

### CODE REVIEW

**Review by dzhulgakov — COMMENTED** [review]
This is fancy :) Is there a measurable impact for compile time? (I'm a bit scared of too many crc compiler computation)

**Comment by smessmer** [comment]
I did not measure but template instantiation rules should make sure that it's only computed once per type. The C++14 computation is an iterative loop and should be very fast, crc is designed to be fast and is a single pass over the data. The C++11 crc computation is recursive, so it might in theory be a little bit slower, but probably also negligible. Also, it's tail recursive, so the compiler might even be able to optimize it into an iterative loop.

**Comment by ezyang** [comment]
Please measure the compile time impact before and after. You can do it on a single stressful file, if you like.

**Comment by ezyang** [comment]
What happens if the crc collides?

**Comment by smessmer** [comment]
@ezyang if crc collides, we will give the same id to two types, which would be an issue. But it's crc64. This is very unlikely to collide, there is some more info [here](https://stackoverflow.com/questions/6025445/how-likely-are-two-blocks-of-data-likely-to-produce-the-same-crc64-value).

If you think collisions are too likely, I could add a runtime mechanism to check for them (i.e. a runtime type registry that remembers type names and corresponding ids), but I think that's probably not worth it as long as we only need it for collision checking.

I don't expect any compile time issues, but I'll do some measurements.

**Comment by smessmer** [comment]
I did some compile time measurements.

#### Setup
- oss build with GCC8 on a devvm in a Python 2.7 virtualenv.
- Command: `python setup.py clean && git clean -xfd && time python setup.py install`

#### Summary
- This PR introduces a build size regression of less than 0.2% (20sec in a 210min build), this is probably within build flakiness.

#### Without this PR in C++11
Average user build time over both runs: 210m19s

```
real	11m5.594s
user	210m1.055s
sys	11m11.972s
```

```
real	10m56.386s
user	210m37.528s
sys	10m45.378s
```
#### With this PR in C++11
Average user build time over both runs: 210m42s

```
real	11m16.041s
user	211m6.071s
sys	11m17.403s
```

```
real	11m13.680s
user	210m18.007s
sys	10m46.595s
```

File: c10/util/TypeIndex.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+#if !defined(__clang__) && !defined(_MSC_VER) && defined(__GNUC__) && \
+    __GNUC__ < 5
+// Getting __PRETTY_FUNCTION__ at compile time only works with GCC >= 5
+#error "You're running a too old version of GCC. We need GCC 5 or later."
+#endif
+
+template <typename T>
+inline C10_HOST_CONSTEXPR uint64_t type_index_impl() noexcept {
+// Idea: __PRETTY_FUNCTION__ (or __FUNCSIG__ on msvc) contains a qualified name
```
Comment (ezyang) [inline c10/util/TypeIndex.h]: I'm not sure how I feel about this lol

Comment (smessmer) [reply c10/util/TypeIndex.h]: It's more reliable and easier to understand/change than the ODR hacks we had before.

**Comment by ezyang** [comment]
I'm not sure how I feel about this... I guess I'd feel better about this if we had some way (even if it's a runtime, test only check) to see if there any conflicts for any of the type ids we use inside our application...

**Comment by smessmer** [comment]
@ezyang there's tests in c10/test/util/TypeIndex_test.cpp

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**